src/multicharge/model/type.F90

```diff
@@ -58,339 +57,342 @@ module multicharge_model_type
-
-   dcn = present(dcndr) .and. present(dcndL)
-   grad = present(gradient) .and. present(sigma) .and. dcn 
-   cpq = present(dqdr) .and. present(dqdL) .and. dcn
-
-   call self%update(mol, grad.or.cpq, cache)
-   !if(allocated(cache%cmat)) then 
-   !   cache%cmat = 0.0_wp
-   !   do iat = 1, mol%nat
-   !      cache%cmat(iat, iat) = 1.0_wp
-   !   end do 
-   !
-   !end if
-   !if(allocated(cache%dcdr)) then 
-   !   cache%dcdr = 0.0_wp
-   !   cache%dcdL = 0.0_wp
-   !end if
-
-   allocate(amat(ndim, ndim), xvec(ndim))
-   allocate(ipiv(ndim))
-   if (grad.or.cpq) then
-      allocate(dxdr(3, mol%nat, ndim), dxdL(3, 3, ndim))
-   end if
-
-   call self%get_vrhs(mol, cache, cn, qloc, xvec, dcndr, dcndL, &
-      & dqlocdr, dqlocdL, dxdr, dxdL)
-
-   if (any(mol%periodic)) then
-      call self%get_amat_3d(mol, cache, wsc, alpha, amat)
-   else
-      call self%get_amat_0d(mol, cache, cn, qloc, amat)
-   end if
-
-   vrhs = xvec
-   ainv = amat
-
-   call sytrf(ainv, ipiv, info=info, uplo='l')
-
-   if (info == 0) then
-      if (cpq) then
-         call sytri(ainv, ipiv, info=info, uplo='l')
-         if (info == 0) then
-            call symv(ainv, xvec, vrhs, uplo='l')
-            do ic = 1, ndim
-               do jc = ic+1, ndim
-                  ainv(ic, jc) = ainv(jc, ic)
+      class(mchrg_model_type), intent(in) :: self
+      type(structure_type), intent(in) :: mol
+      real(wp), intent(inout), contiguous :: cn(:)
+      real(wp), intent(inout), contiguous :: qloc(:)
+      real(wp), intent(in), contiguous, optional :: dcndr(:, :, :)
+      real(wp), intent(in), contiguous, optional :: dcndL(:, :, :)
+      real(wp), intent(in), contiguous, optional :: dqlocdr(:, :, :)
+      real(wp), intent(in), contiguous, optional :: dqlocdL(:, :, :)
+      real(wp), intent(out), contiguous, optional :: qvec(:)
+      real(wp), intent(inout), contiguous, optional :: energy(:)
+      real(wp), intent(inout), contiguous, optional :: gradient(:, :)
+      real(wp), intent(inout), contiguous, optional :: sigma(:, :)
+      real(wp), intent(out), contiguous, optional :: dqdr(:, :, :)
+      real(wp), intent(out), contiguous, optional :: dqdL(:, :, :)
+
+      integer :: ic, jc, iat, ndim
+      logical :: grad, cpq, dcn
+      integer(ik) :: info
+      integer(ik), allocatable :: ipiv(:)
+
+      !> Variables for solving ES equation
+      real(wp), allocatable :: xvec(:), vrhs(:), amat(:, :)
+      real(wp), allocatable :: ainv(:, :)
+      !> Gradients
+      real(wp), allocatable :: dadr(:, :, :), dadL(:, :, :), atrace(:, :)
+      real(wp), allocatable :: dxdr(:, :, :), dxdL(:, :, :)
+      type(cache_container), allocatable :: cache
+      real(wp), allocatable :: trans(:, :)
+
+      !> Calculate gradient if the respective arrays are present
+      dcn = present(dcndr) .and. present(dcndL)
+      grad = present(gradient) .and. present(sigma) .and. dcn
+      cpq = present(dqdr) .and. present(dqdL) .and. dcn
+
+      !> Update cache
```

Comment:
not sure what you're referring to

Reply:
I meant this part: 

"Why are those not inout anymore? Generally, the inout is a bit dangerous since we could modify a CN which a different part of the program calculated. Can we leave everything just intent(in)? Maybe just check if something is passed in but don't return? Or we could setup the cache in main and give for everything the option of passing something to the cache (I think, we have all the necessary information there to. Then it would be three steps new_model -> update cache -> solve"